design/userdata-merging.md

```diff
@@ -0,0 +1,234 @@
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+            * [Story 1](#story-1)
+            * [Story 2](#story-2)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks](#drawbacks-optional)
+      * [Alternatives](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+With v1alpha2, Cluster-API goal is to split the bootstrap provider from the
+infrastructure provider. However, in the case of baremetal, such a split is not
+straightforward. Some configuration is inherently tied to the physical node (for
+example networking, with bonds, vlans etc. or hardware specific setup such as
+drivers etc.). There is a need for giving the user the possibility to provide
+a node-specific userdata. In this proposal we outline a solution that would
+merge different userdata inputs.
+
+## Motivation
+
+### Goals
+
+* Provide support for user-provided cloud-config per node
+* Validate the user-provided cloud-config (to prevent conflict with CABPK
+  output)
+* Merge the inputs from CABPK and user
+* Support later addition of further configurations (Ignition etc.).
+
+### Non-Goals
+
+ * Support other type of userdata, considering that Cluster-api only supports
+   cloud-init based bootstrap provider (CABPK) we will focus on cloud-init only
+   in this proposal.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+Considering a set of motley physical nodes, all with different networking
+setup, the user will be able to specify a per-node network configuration. Hence
+the BaremetalHost selection can be independent, not requiring to tie
+BaremetalHost and BaremetalMachines through labels to give the networking
+through a script passed to CABPK.
+
+#### Story 2
+
+Considering a CAPI deployment, it would be possible to specify some
+node-specific configuration related to the physical node, offering the user the
+possibility to fully use the deployment features such as scale-in/scale-out
+independently. The deployment could then use physical nodes that are not
+strictly identical.
+
+### Implementation Details/Notes/Constraints
+
+This does not require to modify the current CAPI workflow.
+When selecting a baremetalHost for the BaremetalMachine, the controller would
+merge the userdata from the machine with the user-provided input and the
+load-balancer related input. The output would be written as userdata in the
+BaremetalHost for deployment.
```

Comment:
I think I mentioned this previously, but having some hard-coded merge implementation makes me nervous, there are multiple approaches to merging and cloud-init already supports merging multiple configs (e.g multiple cloud-config parts in a mime archive), including specifying the strategy to apply when merging:

https://cloudinit.readthedocs.io/en/latest/topics/merging.html

I get that one of the objectives here is validation, but I wonder if we'll end up reimplementing all that cloud-init merge logic incrementally - maybe it would be better to just consider appending user-data and let the tool take care of the merging?